File: sdks/typescript/src/runtime/call.ts
import type { BaseClient } from "./client";
import type { MethodCall } from "./types";

/**
 * A lazy, awaitable API call. Awaiting it sends the call on its own; passing it
 * to {@link BaseClient.batch} bundles it into one request with typed results.
 */
export class Call<T> implements PromiseLike<T> {
    public constructor(
        private readonly client: BaseClient,
        public readonly methodCall: MethodCall,
        public readonly decode?: (raw: any) => T,
    ) {}

    /** Turn a built response envelope into the typed value for this call. */
    public build(raw: any): T {
        return this.decode ? this.decode(raw) : (raw as T);
    }

    public then<R1 = T, R2 = never>(
        onfulfilled?: ((value: T) => R1 | PromiseLike<R1>) | null,
        onrejected?: ((reason: unknown) => R2 | PromiseLike<R2>) | null,
    ): Promise<R1 | R2> {
        return this.client
            .sendOne(this.methodCall)
            .then((raw) => this.build(raw))
            .then(onfulfilled, onrejected);
    }
}

export type CallResults<C extends readonly Call<any>[]> = {
    [K in keyof C]: C[K] extends Call<infer U> ? U : never;
};

File: sdks/python/src/pecans/runtime/call.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Callable, Generator, Generic, TypeVar

from .types import MethodCall

if TYPE_CHECKING:
    from .client import BaseClient

T = TypeVar("T")


class Call(Generic[T]):
    """A lazy, awaitable API call. Awaiting it sends the call on its own; passing it
    to :meth:`BaseClient.batch` bundles it into one request with typed results."""

    def __init__(
        self,
        client: BaseClient,
        method_call: MethodCall,
        decode: Callable[[Any], T] | None = None,
    ) -> None:
        self.client = client
        self.method_call = method_call
        self.decode = decode

    def build(self, raw: Any) -> T:
        """Turn a built response envelope into the typed value for this call."""
        if self.decode is not None:
            return self.decode(raw)

        return raw

    async def _send(self) -> T:
        raw = await self.client.send_one(self.method_call)

        return self.build(raw)

    def __await__(self) -> Generator[Any, None, T]:
        return self._send().__await__()
